### ISSUE TITLE
Missing Functional Threshold Heart Rate and Best 20min Power

### ISSUE THREAD

**Issue by f1lterr** [thomaschampagne/stravistix#574]
### Step 1: Are you in the right place?

  * Your problem may have been already covered in the FAQ. Did you check https://github.com/thomaschampagne/stravistix/wiki/Frequently-Asked-Questions?
  * Your problem or feature request may already exists in project issues section. Did you check https://github.com/thomaschampagne/stravistix/issues?

### Step 2: Describe your environment

  * Plugin version: 6.1.1 Stable
  * Chrome/Opera/Chromium version: Chrome Version 64.0.3282.186 (Official Build) (64-bit)
  * OS version: Win10
  
### Step 3: Describe the problem:

Hi,

First, thanks for developing Stravistix. :-)

I'm missing Functional Threshold Heart Rate and Best 20min Power in the Stravistix stats panel in Strava.

I'm pretty sure this was working in the past and just stopped. 

Interestingly enough, I do see this stat for some people I'm following. 

Am I missing a setting somewhere? 

I've tried removing and re-installing the plugin to no effect. I've also confirmed the same behavior on a completely different laptop, just to be sure it wasn't something specific to my Chrome installation. 

#### Steps to reproduce:

  1. Look for Functional Threshold Heart Rate and Best 20min Power in the Stravistix stats panel for an activity in Strava. More specifically, my activities. I see these stats for people I'm following.
  
#### Actual Behavior:

  * What happened?
  * Do you have javascript errors logs in browser console?

  > Press F12 to see developer console and javascript logs

  ```javascript
 19:23:01.422 jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22 A parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/?file=gdata&v=1.x, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message. See https://www.chromestatus.com/feature/5718547946799104 for more details.
google.loader.f @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22
P.load @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:30
google.load @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:18
(anonymous) @ 1502864063:3356
19:23:01.423 jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22 A parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/?file=gdata&v=1.x, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message. See https://www.chromestatus.com/feature/5718547946799104 for more details.
google.loader.f @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22
P.load @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:30
google.load @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:18
(anonymous) @ 1502864063:3356
19:23:01.446 jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22 A parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/api/gdata/1.10/208a907929e7936990d2bf48205aa8d8/core,mediarss,geo,opensearch,atom,threading,app,gdata,finance,contacts,analytics,blogger,maps,gbase,acl,calendar,default.I.js, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message. See https://www.chromestatus.com/feature/5718547946799104 for more details.
google.loader.f @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22
(anonymous) @ ?file=gdata&v=1.x:8
19:23:01.459 jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22 A parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/api/gdata/1.10/208a907929e7936990d2bf48205aa8d8/core,mediarss,geo,opensearch,atom,threading,app,gdata,finance,contacts,analytics,blogger,maps,gbase,acl,calendar,default.I.js, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message. See https://www.chromestatus.com/feature/5718547946799104 for more details.
google.loader.f @ jsapi?key=ABQIAAAA9S76sdh5KSPpfr65zxQHGBT-rU0CFRjazioGWeeHjJLEyYXO2RSmaFbEdGganPrQRY2l3ORmYrTPpA:22
(anonymous) @ ?file=gdata&v=1.x:8
19:23:04.119 ActivityProcessor.js:39 Using existing activity cache mode
19:23:09.126 StravistiX.js:724 A previous sync exists on Thu Apr 12 2018 19:22:34 GMT-0400 (Eastern Daylight Time)
19:23:09.137 ActivitiesSynchronizer.js:398 Fast sync mode enabled
19:23:09.509 util.js:228 Google Maps API warning: NoApiKeys https://developers.google.com/maps/documentation/javascript/error-messages#no-api-keys
yw.j @ util.js:228
(anonymous) @ js?sensor=false:152
(anonymous) @ js?sensor=false:62
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:62
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
Md @ js?sensor=false:62
Ld.xa @ js?sensor=false:130
(anonymous) @ js?sensor=false:30
(anonymous) @ common.js:1
19:23:09.510 util.js:228 Google Maps API warning: SensorNotRequired https://developers.google.com/maps/documentation/javascript/error-messages#sensor-not-required
yw.j @ util.js:228
(anonymous) @ js?sensor=false:152
(anonymous) @ js?sensor=false:62
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:62
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
(anonymous) @ js?sensor=false:60
(anonymous) @ js?sensor=false:130
Md @ js?sensor=false:62
Ld.xa @ js?sensor=false:130
(anonymous) @ js?sensor=false:30
(anonymous) @ common.js:1
19:23:09.562 ActivitiesSynchronizer.js:408 Local and remote activities count matches.
19:23:09.563 ActivitiesSynchronizer.js:418 Skipping HistoryChanges analisys
19:23:09.844 ActivitiesSynchronizer.js:450 Last sync date time saved:  Thu Apr 12 2018 19:23:09 GMT-0400 (Eastern Daylight Time)
19:23:10.201 ActivitiesSynchronizer.js:460 Sync With Athlete Profile done
19:23:10.202 StravistiX.js:732 Fast sync finished {globalHistoryChanges: {…}, computedActivities: Array(504), lastSyncDateTime: 1523575389564, syncWithAthleteProfile: {…}}

  ```
  
#### Expected Behavior:

  * What did you expect to happen?
  
#### Media attachement:

![image](https://user-images.githubusercontent.com/38335300/38709149-70efd24a-3e87-11e8-87b8-21a7e5bfcfbb.png)

>You can upload/attach any images to this ticket.

#### Providing your stravistix synced athlete history can be usefull to solve the problem

> To export you history click on "sync button" (top right and corner), then "Backup history". You can zip the file and upload it here. 

[2018.4.12-19.27_v6.1.1.history.zip](https://github.com/thomaschampagne/stravistix/files/1905306/2018.4.12-19.27_v6.1.1.history.zip)

**Comment by f1lterr** [comment]
Hi Thomas,

I'm pretty sure this is a case of PEBKAC and suspect you'll reply with "did you check xyz?" and I will have an aha moment.

Thanks!

-Steve

**Comment by thomaschampagne** [comment]
Hi

Do you have an activity id?
Did your activity paused a long time? If paused more than 3000 sec then calculation is skipped https://github.com/thomaschampagne/stravistix/blob/b23cfdd71aa3dec873283762f3806b06138752e2/plugin/common/scripts/SplitCalculator.ts#L38

I let you close the ticket if that's the case :)

Tom

**Comment by f1lterr** [comment]
Hmmm…. No long pauses that I’m aware of. None of my activities show the data in question.

 

Here are some recent IDs if you want to check them out:

 

https://www.strava.com/activities/1504565553

https://www.strava.com/activities/1502864063

https://www.strava.com/activities/1498927438

https://www.strava.com/activities/1484466098

https://www.strava.com/activities/1464376248

 

I wish I knew for sure if the “Best 20min Power” stat ever worked. I want to say I’m 99% positive it did back when it was called “FTP” in the previous plugin version.

 

Let me know if you need anything from me – I’m more than happy to put in some effort.

 

Thanks!

 

-Steve

**Comment by thomaschampagne** [comment]
Your ticket has been closed :) . Could test/validate that everything is OK using latest generated build from https://bitbucket.org/thomaschampagne/stravistix-ci-builds/downloads/ ?

If you don't approve the test build, you can reopen this ticket :)

**Comment by f1lterr** [comment]
Hmmm…

 

This build, right? 

 

https://bitbucket.org/thomaschampagne/stravistix-ci-builds/downloads/v6.1.1_stable_2018-04-13-10-12.develop.aa42b2f6.zip

 

Still seeing the issue on two different machines. 

 

https://www.strava.com/activities/1504565553

**Comment by f1lterr** [comment]
Sorry Thomas… false alarm, I think.

 

After a few cycles of clearing all data and resyncing, I’m pretty sure I’m good. I’ll let you know if that’s not the case.

 

Thanks so much for the support!

**Comment by thomaschampagne** [comment]
That's the good build. It seems the build use cache generated by current 6.1.1

In the App:
- Go to Advanced menu
- Click "Clear the plugin cache" button
- Reload your activity

The cause is that your Bkool activities have the time not scaled properly. You should get time (seconds) samples like: 1,2,3,5,6,9,10,......

And your bktool ones have sometime 1,2,3,**5,5,5**,10,... meaning that you sample 3 times  on same second

Plugin tries to linear interpolate missing value between 5 - 10. Does nothing between 1 - 2 - 3. And don't know what to do between 5 - 5 - 5.

**Comment by thomaschampagne** [comment]
Ok seen your last message after my last one

**Comment by f1lterr** [comment]
Hi Thomas,

 

My apologies if this email reopens the ticket. If that’s the case, feel free to re-close or I can do it.

 

The “nuclear option” of removing all Stravistix plugins and starting fresh with the latest build seems to have addressed my issues. :-) 

 

I just had to thank you for the detailed info and quick response. I continue to be amazed by people like you and Eric at Fitnesssyncer and Ben at Veloviewer  who provide support that far exceeds most major, commercial software vendors. That’s why I always contribute/subscribe. 

 

Cheers from Trumpland. :|

 

-Steve
